frontend/packages/console-app/src/components/user-preferences/language/LanguageDropdown.tsx

```diff
@@ -0,0 +1,87 @@
+import * as React from 'react';
+import { QuickStartContext, QuickStartContextValues } from '@patternfly/quickstarts';
+import { Skeleton, SelectOption, Select, SelectVariant } from '@patternfly/react-core';
+import { useTranslation } from 'react-i18next';
+import { getProcessedResourceBundle } from '../../quick-starts/utils/quick-start-context';
+import { supportedLocales } from './const';
+import { usePreferredLanguage } from './usePreferredLanguage';
+
+const LanguageDropdown: React.FC = () => {
+  // resources and calls to hooks
+  const { i18n, t } = useTranslation();
+  const { setResourceBundle } = React.useContext<QuickStartContextValues>(QuickStartContext);
+  const languageOptions = Object.keys(supportedLocales).map((lang) => ({
+    lang,
+    label: supportedLocales[lang],
+  }));
+  const [preferredLanguage, setPreferredLanguage, preferredLanguageLoaded] = usePreferredLanguage();
+  const [dropdownOpen, setDropdownOpen] = React.useState(false);
+  const loaded: boolean = preferredLanguageLoaded;
+
+  const selectOptions: JSX.Element[] = React.useMemo(
+    () =>
+      languageOptions.map((langOption) => (
+        <SelectOption key={langOption.lang} value={langOption.label} />
+      )),
+    [languageOptions],
+  );
+
+  const initialLanguage =
+    preferredLanguage ||
+    // handles languages we support, languages we don't support, and subsets of languages we support (such as en-us, zh-cn, etc.)
+    i18n.languages.find((lang) => languageOptions.some((langOption) => langOption.lang === lang));
+
+  // utils and callbacks
+  const getDropdownLabelForValue = (): string =>
+    languageOptions.find((langOption) => langOption.lang === initialLanguage)?.label;
+  const getDropdownValueForLabel = (selectedLabel: string): string =>
+    languageOptions.find((langOption) => langOption.label === selectedLabel)?.lang;
+  const onToggle = (isOpen: boolean) => setDropdownOpen(isOpen);
+  const setLanguage = (language: string) => {
+    i18n.changeLanguage(language || 'en');
+    setPreferredLanguage(language);
+  };
+  const onSelect = (_, selection) => {
+    const selectedValue = getDropdownValueForLabel(selection);
+    selectedValue !== preferredLanguage && setLanguage(selectedValue);
+    setDropdownOpen(false);
+  };
```

Comment:
You can keep `setLanguage` if you want. But I don't see a reason for not in-lining this if there is only one code area which calls `setLanguage`. With the change to the SelectOption the selection is already the right language (key).

```suggestion
  const onSelect = (_, selectedLang: string) => {
    if (selectedLang !== preferredLanguage) {
      i18n.changeLanguage(selectedLang || 'en');
      setPreferredLanguage(selectedLang);
    }
    setDropdownOpen(false);
  };
```

Reply:
removed `setLanguage` to inline calls